SeaPublicWebsite/Controllers/EnergyEfficiencyController.cs

```diff
@@ -1,1338 +1,1515 @@
+            };
+            return View("FindEpc", viewModel);
+        }
+
+        [HttpPost("find-epc/{reference}")]
+        public async Task<IActionResult> FindEpc_Post(FindEpcViewModel viewModel)
+        {
+            if (!ModelState.IsValid)
+            {
+                return await FindEpc_Get(viewModel.Reference);
+            }
+
+            var propertyData = await propertyDataStore.LoadPropertyDataAsync(viewModel.Reference);
+            
+            propertyData.FindEpc = viewModel.FindEpc;
+            propertyData.EpcAddressConfirmed = null;
+            propertyData.EpcDetailsConfirmed = null;
+            propertyData.EpcCount = null;
+            propertyData.Epc = null;
+            propertyData.PropertyType = null;
+            propertyData.YearBuilt = null;
+            PropertyDataHelper.ResetUnusedFields(propertyData);
+            await propertyDataStore.SavePropertyDataAsync(propertyData);
+
+            var forwardArgs =
+                questionFlowService.ForwardLinkArguments(QuestionFlowPage.FindEpc, propertyData, viewModel.EntryPoint);
+            return RedirectToAction(forwardArgs.Action, forwardArgs.Controller, forwardArgs.Values);
+        }
+
+        [HttpGet("postcode/{reference}")]
+        public async Task<IActionResult> AskForPostcode_Get(string reference)
+        {
+            var propertyData = await propertyDataStore.LoadPropertyDataAsync(reference);
+            
+            var backArgs = questionFlowService.BackLinkArguments(QuestionFlowPage.AskForPostcode, propertyData);
+            var skipArgs = questionFlowService.SkipLinkArguments(QuestionFlowPage.AskForPostcode, propertyData);
+            var viewModel = new AskForPostcodeViewModel
+            {
+                Postcode = propertyData.Postcode,
+                HouseNameOrNumber = propertyData.HouseNameOrNumber,
+                Reference = propertyData.Reference,
+                BackLink = Url.Action(backArgs.Action, backArgs.Controller, backArgs.Values),
+                SkipLink = Url.Action(skipArgs.Action, skipArgs.Controller, skipArgs.Values)
+            };
+
+            return View("AskForPostcode", viewModel);
+        }
+
+        [HttpPost("postcode/{reference}")]
+        public async Task<IActionResult> AskForPostcode_Post(AskForPostcodeViewModel viewModel)
+        {
+            if (viewModel.Postcode is not null && !(await postcodesIoApi.IsValidPostcode(viewModel.Postcode)))
+            {
+                ModelState.AddModelError(nameof(AskForPostcodeViewModel.Postcode), "Enter a valid postcode in the format AB12 3CD");
+            }
+            
+            if (!ModelState.IsValid)
+            {
+                return await AskForPostcode_Get(viewModel.Reference);
+            }
+            List<EpcInformation> epcInformationList = await epcApi.GetEpcsInformationForPostcodeAndBuildingNameOrNumber(viewModel.Postcode, viewModel.HouseNameOrNumber);
```

Comment:
Do we really need to do this here? It's an extra API call

Reply:
The way my questionflow logic is set up, yes, because we need to rediret to a separate NoEpcFound page if epcCount is 0 so we need to run the call before reaching ConfirmAddress, and then again if we get there.